numpy/random/mtrand/mtrand.pyx

```diff
@@ -5064,7 +5064,8 @@ cdef class RandomState:
             x_ptr = <char*><size_t>x.ctypes.data
             stride = x.strides[0]
             itemsize = x.dtype.itemsize
-            buf = np.empty_like(x[0])  # GC'd at function exit
+            # Ensure that buf has the same dtype as x
+            buf = np.empty_like(x[:1])  # GC'd at function exit
```

Comment:
@simongibbons is there a bug, can't figure it out right now? For the multi-dimensional case, nothing should change as far as I can see. For the 1D case, it seems to me things may be broken if the dtype of the result of `arr[0]` (the scalar) is smaller then an object pointer, the buffer is too small, but this change should fix that as well.